### ISSUE TITLE
Add a way to display the mouse and lock keyboard input.

### ISSUE THREAD

**Issue by etra0** [veeenu/hudhook#36]
I'm still not totally sure about this but it would be worth the discussion.

Currently, although this captures mouse input, it doesn't have any mechanism to 'force' the cursor to be displayed.
I'd like to be something like Reshade, when the Menu is open, show and capture mouse input (and blocks mouse input
from the game).

Right now I'm doing some *very* nasty hack for the tool I've been building for The Witcher and it kinda works,
it shows the mouse even if the game doesn't want to:

https://github.com/etra0/hudhook/compare/master...temp_patches

This could work as a proof of concept I guess if we test it in more games.

**Comment by veeenu** [comment]
I used to have something similar way back when and I think it would be a good idea, because some games will definitely need it.

I don't think your solution is particularly hacky, it's definitely the "intended" way of showing a cursor. The main thing I'd change is that the decision to show the cursor should be delegated to the clients via some API. I'm not sure what that could look like yet.

**Comment by etra0** [comment]
The solution is actually using [`ImguiRenderLoop::should_block_messages`](https://github.com/veeenu/hudhook/blob/d8f18f4dd3c0535d459f104527ff705f849b6b54/src/hooks/common.rs#L181) so it's up to the client to re-implement that trait in order to show the cursor. Maybe could be renamed

**Comment by etra0** [comment]
I'd also like a way for the mouse input to be totally blocked, that would require a bit of investigation, because sometimes when you click you also interact with the game and it's a bit messy. Right now my hack only displays the cursor

**Comment by veeenu** [comment]
I've kind of given up on that. Been trying to block input for DS3 and ER for _years_ now, to absolutely no avail. Still haven't understood how to block their input flow, I've no idea where they even get the data from.

**Comment by etra0** [comment]
Good thing is reshade is also open source :D I'll try some stuff when I have the time and see if a PR is possible

**Comment by Godnoken** [comment]
I'm also very interested in finding a way to block game input completely, in all games. I'm going to investigate it over the next month if time allows.

> Good thing is reshade is also open source :D I'll try some stuff when I have the time and see if a PR is possible

Are you saying that Reshade implements this behaviour?

**Comment by etra0** [comment]
iirc yes it does, but I may be wrong. It's been a while since I had time to look this up.

**Comment by Godnoken** [comment]
Interesting. I'm going to give it a download and see if it works for a wide variety of games.

**Comment by Godnoken** [comment]
It seems to work almost flawlessly, amazing.

The only game I had a problem with so far is Stardew Valley. While game input is definitely blocked, the game 'drops' the current mouse position/inactivates when the menu comes up. Meaning - hovering elements stop showing.
All other games I tried did just fine on that part.

One other issue that isn't huge is that the in-game cursor (only the visuals) follow the hardware cursor in the Reshade menu in some games. However, game input is still blocked.

Now the question is how this magic works.. I'll update if I manage to get something working.

Could you possible try Reshade for DS3 or ER and confirm that mouse input gets blocked @veeenu?

**Comment by veeenu** [comment]
I can confirm the input gets blocked for Elden Ring! Haven't tried for DS3 but I'm ready to bet it is the same there. Now to understand how it works... 😅

**Comment by Godnoken** [comment]
Awesome. Yeah, there's a lot of interesting stuff in Reshade. The installation process is one.

Anyways - I've never done C++ so I haven't got very far with trying to understand the game input block nor how it manages to display the second cursor..

Have you had any time to look at it?

**Comment by etra0** [comment]
FWIW I have some (a bit dirty) patches to force to show the cursor, but I haven't blocked the input yet, you can check them in https://github.com/etra0/hudhook/commits/temp_patches

**Comment by Godnoken** [comment]
> Unfortunately not. Been pretty busy lately, had to cut back a bit on working on these things 😞

No worries, I'll keep jabbing at it when I get the time too.


> FWIW I have some (a bit dirty) patches to force to show the cursor, but I haven't blocked the input yet, you can check them in https://github.com/etra0/hudhook/commits/temp_patches

I actually tried your solution but I couldn't stop the mouse from flickering constantly, if I remember correctly.. Any ideas?
May have another go at it now.

**Comment by etra0** [comment]
hmm, could it be because at every iteration I'm doing ShowCursor. I don't remember if I just do it once it'll show... I can check that during the weekend.

**Comment by Godnoken** [comment]
Well, I tried it yesterday and while the cursor indeed did show up, it did not do so consistently - at the very least not on my end.
So the cursor would show up when constructed to do so (initially), but whenever it is constructed to not show up, it doesn't seem to obey. The cursor disappears but after a while it pops back up without the ShowCursor being triggered. It may be an issue from game to game, not sure.

**Comment by Godnoken** [comment]
Got an update on this. I've managed to block game input completely similarly to how Reshade has done it. It doesn't work *fully* in all games (just like Reshade).

- As far as I have tested, all clicks or keypresses will be ignored by the game. - Won't work if the game uses its own input sys or dinput
- In some games, hovering elements are still not ignored - Reshade seems to manage this flawlessly, not sure how
- In some games, the in-game cursor's movement gets ignored now (ideal) - However, Reshade hasn't figured out how to do this globally either, and per my estimations, it seems impossible to find a global solution..

![hudhook_blocked_input](https://user-images.githubusercontent.com/17499594/233839376-9137880d-60dd-4f54-8776-f1ae58fa2bb3.gif)

The issue we seem to have with the current WndProc solution is that it only gets hooked into one of the game's threads (probably always the main thread..?), which in some cases makes us miss out on messages that are handled in different threads. 

I've fixed this by utilizing [SetWindowsHookEx](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowshookexa) with [WH_GETMESSAGE](https://learn.microsoft.com/en-us/windows/win32/winmsg/about-hooks#wh_getmessage) and [GetMsgProc](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms644981(v=vs.85)) to hook into all threads inside of the game.

However, I may be wrong and that is entirely unnecessary. As far as I can tell, Reshade only hooks ``GetMessage`` and ``PeekMessage`` once and it still manages to block input like I have done. Either Reshade finds the correct thread inside the game or it is indeed hooked onto all threads as I have done.

Excuse my horribly chaotic code, but [here's](https://github.com/Godnoken/hudhook/tree/revamp-input-handling) a working branch.
You can ignore all hooks apart from get_msg_proc as the others are currently not a part of the solution.

Let me know if you test it, intrigued to see if it works in Elden Ring or DS3 too.

**Comment by Godnoken** [comment]
Forgot to mention, the game input issues I have mentioned earlier in [disccussions](https://github.com/veeenu/hudhook/discussions/56) are completely gone now, so that's another positive side-effect of this.

**Comment by veeenu** [comment]
That is absolutely awesome, nice find!

I'm a bit skeptical of adding a hook kind as a solution, it seems complicated and disruptive to the API. But if we manage to use `SetWindowsHookEx` in the initialization/deinitialization of a renderer object so that it can manage the resulting state, that should be ideal, provided it doesn't add a lot of overhead.

**Comment by veeenu** [comment]
> Hmm, with "hook kind", do you mean the hooks in `CommonHooks` or `SetWindowsHookEx`? And the "resulting state" is referring to what exactly?

Yep, that one. I wouldn't add that, I'd rather have the setup/teardown of `SetWindowsHookEx` be done by the renderers themselves. Much like in the same spots the setting/unsetting of the wnd proc is done now.
 
> It's definitely implemented in a really ugly way right now. I'm still too inexperienced and I've been too lazy to learn what's idiomatic and generally just go with whatever seems to work, hehe.
> 
> Shall I create a pull request where we could work on this?

Oh sure, feel free to open one! I'd start from a fresh slate off of `main` though, so that we can end up with a small enough PR diff.

**Comment by Godnoken** [comment]
> Yep, that one. I wouldn't add that, I'd rather have the setup/teardown of `SetWindowsHookEx` be done by the renderers themselves. Much like in the same spots the setting/unsetting of the wnd proc is done now.
> By "resulting state" I mean the state of the message hook being active/inactive and the consequent change of flow in the application.

Rightio, I'll look into it!

> Oh sure, feel free to open one! I'd start from a fresh slate off of `main` though, so that we can end up with a small enough PR diff.

Cool, I'll most likely get on it first thing in the morning.

**Comment by etra0** [comment]
This sound like some exciting progress! I'll check your branch in more depth as soon as I have time, as my tool could also benefit from this!

**Comment by Godnoken** [comment]
There is most definitely a major fuck up somewhere with the input. At some point during `should_block_messages`, the input freezes completely in Fallout 3 and in other games I've had the in-game mouse just locking up. ALT + TAB *sometimes* fixes it.

I'm an übernoob when it comes to threads, pointers, mutexes etc.. I'll get the PR up tomorrow so if anyone wants to dabble with it, it'll be much appreciated :)

**Comment by etra0** [comment]
hey, any reason we can dig back at this? maybe just first focus on getting the cursor stuff. I'm interested to know if such feature is valuable to see if I can take what's been done and 'formalize' it.

I'm trying to avoid to use my own personal hudhook fork to be more 'standard' haha.
